src/ens/records.c

```diff
@@ -0,0 +1,164 @@
+ *
+ * @param record pointer to the location, where the loaded record shall be stored
+ * @param target timestamp for which to find the nearest entry for
+ * @param start lower bound for the binary search
+ * @param end upper bound for the binary search
+ */
+int find_record_via_binary_search(record_t* record,
+                                  uint32_t target,
+                                  record_sequence_number_t start,
+                                  record_sequence_number_t end) {
+    record_t start_record;
+    record_t end_record;
+
+    // load the initial start and end record
+    int rc = load_contact(&start_record, start);
+    if (rc) {
+        return rc;
+    }
+    rc = load_contact(&end_record, end);
+    if (rc) {
+        return rc;
+    }
+
+    do {
+        // calculate the contact in the middle between start and end and load it
+        record_sequence_number_t middle = (start_record.sn + end_record.sn) / 2;
+        int rc = load_contact(record, middle);
+        if (rc) {
+            return rc;
+        }
+
+        // determine the new start and end
+        if (record->timestamp > target) {
+            memcpy(&start_record, record, sizeof(record_t));
+        } else {
+            memcpy(&end_record, record, sizeof(record_t));
+        }
+
+        // break, if we are at the exact timestamp or our start and end are next to each other
+    } while (record->timestamp != target && (end_record.sn - start_record.sn) > 1);
+
+    return 0;
+}
+
+int ens_records_iterator_init_timerange(record_iterator_t* iterator, uint32_t* ts_start, uint32_t* ts_end) {
+    record_sequence_number_t oldest_sn = get_oldest_sequence_number();
+    record_sequence_number_t latest_sn = get_latest_sequence_number();
+
+    // try to find the oldest contact in our timerange
+    record_t start_rec;
+    int rc = load_contact(&start_rec, oldest_sn);
+    if (rc) {
+        return rc;
+    }
+    // if starting timestamp lies in our bounds, perform binary search
+    if (start_rec.timestamp < *ts_start) {
+        rc = find_record_via_binary_search(&start_rec, *ts_start, oldest_sn, latest_sn);
+
+        if (rc) {
+            return rc;
+        }
+    }
+
+    // try to find the newest contact within out timerange
+    record_t end_rec;
+    rc = load_contact(&end_rec, latest_sn);
+    if (rc) {
+        return rc;
+    }
+    // if ending timestamp lies in our bounds, perform binary search
+    if (end_rec.timestamp > *ts_end) {
+        rc = find_record_via_binary_search(&end_rec, *ts_end, oldest_sn, latest_sn);
+
+        if (rc) {
+            return rc;
+        }
+    }
+
+    iterator->finished = false;
+    iterator->sn_next = start_rec.sn;
+    iterator->sn_end = end_rec.sn;
+
+    return 0;
+}
+
+record_t* ens_records_iterator_next(record_iterator_t* iter) {
+    if (iter->finished) {
+        return NULL;
+    }
+
+    record_t* next = NULL;
+
+    // What is this?
+    while (next == NULL) {
```

Comment:
This will block in case we reached the last record but could not load it.
We also need to check that the iter->finished == false.
```suggestion
    while (next == NULL && !iter->finished) {
```